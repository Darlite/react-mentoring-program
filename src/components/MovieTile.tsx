import styles from '../styles/MovieTile.module.css';
import React, {useState} from "react";
import {MovieDetailsData} from "../types/MovieDetailsData";

interface MovieTileProps {
    movieDetails: MovieDetailsData,
    onClick: (movieDetails:MovieDetailsData ) => void;
}

const MovieTile: React.FC<MovieTileProps> = ({movieDetails, onClick}) => {
    const { imageUrl, movieName, releaseYear, relevantGenres } = movieDetails;

    const [isHovered, setHovered] = useState(false);
    const [showContextMenu, setShowContextMenu] = useState(false);

    function showPopUp() {
        setShowContextMenu(true);
    }

    function closePopUp() {
        setShowContextMenu(false);
    }

    return (
        <div className={styles.movieTile}
             data-testid="movieTile"
             onMouseEnter={() => setHovered(true)}
             onMouseLeave={() => setHovered(false)}>
            <img
                 src={imageUrl}
                 alt={movieName}
                 onClick={() => onClick(movieDetails)} />
            {isHovered && (
                <span className={styles.kebabMenu}
                      role="button"
                      onClick={showPopUp}
                >ï¸™</span>
            )}
            {showContextMenu && (
                <div className={styles.contextMenu} data-testid="contextMenu">
                    <div className={styles.contextMenuClose} onClick={closePopUp}>X</div>
                    <div className={styles.contextMenuOption}>Edit</div>
                    <div className={styles.contextMenuOption}>Delete</div>
                </div>
            )}
            <span className={styles.movieTitleAndYear}>
                <h2 className={styles.title}>{movieName}</h2>
                <span className={styles.year}>{releaseYear}</span>
            </span>
            <p className={styles.genres}>
                {relevantGenres.map(genre => (
                    <span key={Math.random()}>{genre} </span>
                ))}
            </p>
        </div>
    )
}

export default MovieTile;
