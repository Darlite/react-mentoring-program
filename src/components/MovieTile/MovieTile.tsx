import styles from './MovieTile.module.css';
import React, {useState} from "react";
import {MovieDetailsData} from "../../types/MovieDetailsData";
import {GenreType} from "../../constants/GenreType";
import placeholderImage from "../../assets/images/placeholderPoster.jpg"

interface MovieTileProps {
    movieDetails: MovieDetailsData,
    onClick: (movieDetails:MovieDetailsData ) => void;
    handleEdit: () => void;
    handleDelete: () => void;
}

const handleImageError = (e :  React.SyntheticEvent<HTMLImageElement, Event>) => {
    e.currentTarget.src = placeholderImage;
    e.currentTarget.onerror = null;
}

const MovieTile: React.FC<MovieTileProps> = ({movieDetails, onClick, handleEdit, handleDelete}) => {
    const { poster_path, title, release_date, genres } = movieDetails;

    const [isHovered, setHovered] = useState(false);
    const [showContextMenu, setShowContextMenu] = useState(false);

    function showPopUp() {
        setShowContextMenu(true);
    }

    function closePopUp() {
        setShowContextMenu(false);
    }

    return (
        <div className={styles.movieTile}
             data-testid="movie-tile"
             onMouseEnter={() => setHovered(true)}
             onMouseLeave={() => setHovered(false)}>
            <img className={styles.movieTileImage}
                 src={!poster_path ? placeholderImage : poster_path}
                 alt={title}
                 onClick={() => onClick(movieDetails)}
                 onError={handleImageError}
            />
            {isHovered && (
                <span className={styles.kebabMenu}
                      role="button"
                      onClick={showPopUp}
                >ï¸™</span>
            )}
            {showContextMenu && (
                <div className={styles.contextMenu} data-testid="context-menu">
                    <div className={styles.contextMenuClose} onClick={closePopUp}>X</div>
                    <div className={styles.contextMenuOption}
                         onClick={handleEdit}>Edit</div>
                    <div className={styles.contextMenuOption}
                         onClick={handleDelete}>Delete</div>
                </div>
            )}
            <span className={styles.movieTitleAndYear}>
                <h2 className={styles.title}>{title}</h2>
                <span className={styles.year}>{release_date.slice(0, 4)}</span>
            </span>
            <p className={styles.genres}>
                {genres.map((genre: GenreType) => (
                    <span key={genre}>{genre} </span>
                ))}
            </p>
        </div>
    )
}

export default MovieTile;
