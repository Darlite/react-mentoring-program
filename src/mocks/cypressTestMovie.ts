import {GenreType} from "../constants/GenreType";

export const cypressTestMovie = {
    title: "Cypress Test Movie",
    editedMovieTitle: "Cypress Edited Test Movie",
    release_date: "2025-05-04",
    poster_path: "https://www.movieposters.com/cdn/shop/files/tron-ares_avxg7mih_480x.progressive.jpg",
    genres: [GenreType.Action, GenreType.ScienceFiction],
    runtime: "138",
    vote_average: "7.2",
    overview: "A highly sophisticated Program, Ares, is sent from the digital world into the real world on a dangerous mission.",
}
