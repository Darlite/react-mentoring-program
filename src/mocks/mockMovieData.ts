import {GenreType} from "../constants/GenreType";

export const moviesData = {
    id: 1,
    imageUrl: "images/Pulp_Fiction.png",
    title: "Pulp Fiction",
    releaseDate: "1994-09-23",
    movieUrl: "https://www.imdb.com/title/tt0110912",
    genres: [GenreType.Crime, GenreType.Thriller],
    runtime: "2h 34m",
    rating: "8.9",
    description: "Jules Winnfield (Samuel L. Jackson) and Vincent Vega (John Travolta) are two hit men who are out to retrieve a suitcase stolen from their employer, mob boss Marsellus Wallace (Ving Rhames). Wallace has also asked Vincent to take his wife Mia (Uma Thurman) out a few days later when Wallace himself will be out of town. Butch Coolidge (Bruce Willis) is an aging boxer who is paid by Wallace to lose his fight. The lives of these seemingly unrelated people are woven together comprising of a series of funny, bizarre and uncalled-for incidents.â€”Soumitra",
}
