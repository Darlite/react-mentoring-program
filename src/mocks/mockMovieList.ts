import {GenreType} from "../constants/GenreType";

export const mockMovieList = [
    {
        id: 1,
        poster_path: "images/Pulp_Fiction.png",
        title: "Pulp Fiction",
        release_date: "1994-09-23",
        movieUrl: "https://www.imdb.com/title/tt0110912",
        genres: [GenreType.Crime, GenreType.Thriller],
        runtime: "2h 34m",
        vote_average: "8.9",
        overview: "Jules Winnfield (Samuel L. Jackson) and Vincent Vega (John Travolta) are two hit men who are out to retrieve a suitcase stolen from their employer, mob boss Marsellus Wallace (Ving Rhames). Wallace has also asked Vincent to take his wife Mia (Uma Thurman) out a few days later when Wallace himself will be out of town. Butch Coolidge (Bruce Willis) is an aging boxer who is paid by Wallace to lose his fight. The lives of these seemingly unrelated people are woven together comprising of a series of funny, bizarre and uncalled-for incidents.â€”Soumitra",
    },
    {
        id: 2,
        poster_path: "images/Bohemian_Rhapsody.png",
        title: "Bohemian Rhapsody",
        release_date: "2018-10-30",
        movieUrl: "https://www.imdb.com/title/tt1727824",
        genres: [GenreType.Musical, GenreType.Documentary],
        runtime: "2h 15m",
        vote_average: "8.5",
        overview: "With his impeccable vocal abilities, Freddie Mercury and his rock band, Queen, achieve superstardom. However, amidst his skyrocketing success, he grapples with his ego, sexuality and a fatal illness.",
    },
    {
        id: 3,
        poster_path: "images/Kill_Bill_Volume_2.png",
        title: "Kill Bill: Vol. 2",
        release_date: "2004-04-08",
        movieUrl: "https://www.imdb.com/title/tt0378194",
        genres: [GenreType.Action, GenreType.Thriller],
        runtime: "2h 17m",
        vote_average: "8.9",
        overview: "A pregnant woman, codenamed the Bride, sets out on a journey to kill her ex-boss, Bill, and targets his brother, Budd, and Elle Driver, the only two survivors of the Deadly Vipers Assassination Squad.",
    },
]