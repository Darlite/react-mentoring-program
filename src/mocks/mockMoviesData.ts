import {GenreType} from "../constants/GenreType";

export const mockMoviesData = {
    id: 680,
    poster_path: "https://www.movieposters.com/cdn/shop/files/PulpFiction.Ver1_480x.progressive.jpg?v=1741192018",
    title: "Pulp Fiction",
    release_date: "1994-09-10",
    genres: [GenreType.Crime, GenreType.Thriller],
    runtime: "167",
    vote_average: "8.3",
    overview: "Jules Winnfield (Samuel L. Jackson) and Vincent Vega (John Travolta) are two hit men who are out to retrieve a suitcase stolen from their employer, mob boss Marsellus Wallace (Ving Rhames). Wallace has also asked Vincent to take his wife Mia (Uma Thurman) out a few days later when Wallace himself will be out of town. Butch Coolidge (Bruce Willis) is an aging boxer who is paid by Wallace to lose his fight. The lives of these seemingly unrelated people are woven together comprising of a series of funny, bizarre and uncalled-for incidents.â€”Soumitra",
}
