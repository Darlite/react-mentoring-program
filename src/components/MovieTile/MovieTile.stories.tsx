import { Meta, StoryObj } from "@storybook/react";
import MovieTile from "./MovieTile";
import {fn} from "@storybook/test";
import {moviesData} from "../../mocks/mockMovieData";


const meta = {
    component: MovieTile,
    title: 'MovieTile',
    tags: ['autodocs'],
    //ðŸ‘‡ Our exports that end in "Data" are not stories.
    excludeStories: /.*Data$/,
    args: {},
} satisfies Meta<typeof MovieTile>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        movieDetails: moviesData,
        onClick: fn(),
    },
};
