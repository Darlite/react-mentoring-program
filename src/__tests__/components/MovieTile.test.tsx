import MovieTile from "../../components/MovieTile/MovieTile";
import React from "react";
import {fireEvent, render, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {mockMoviesData} from "../../mocks/mockMoviesData";
import placeholderImage from "../../assets/images/placeholderPoster.jpg";

describe("MovieTile", () => {
    it("renders correctly", () => {
        const { container } = render(<MovieTile movieDetails={mockMoviesData}
                                                onClick={() => {}}
                                                handleDelete={() => {}}
                                                handleEdit={() => {}}
        />);
        expect(container).toMatchSnapshot();
    });

    it("shows the kebab menu on hover", () => {
        render(<MovieTile
            movieDetails={mockMoviesData}
            onClick={() => {}}
            handleDelete={() => {}}
            handleEdit={() => {}} />);
        const movieTile = screen.getByTestId("movie-tile");
        userEvent.hover(movieTile);
        expect(screen.getByText("︙")).toBeInTheDocument();
        userEvent.unhover(movieTile);
        expect(screen.queryByText("︙")).not.toBeInTheDocument();
    });

    it("shows and closes the context menu", () => {
        render(<MovieTile movieDetails={mockMoviesData}
                          onClick={() => {}}
                          handleDelete={() => {}}
                          handleEdit={() => {}} />);
        const movieTile = screen.getByTestId("movie-tile");
        userEvent.hover(movieTile);

        const kebabMenu = screen.getByText("︙");
        userEvent.click(kebabMenu);
        expect(screen.getByTestId("context-menu")).toBeInTheDocument();
        expect(screen.getByText("Edit")).toBeInTheDocument();
        expect(screen.getByText("Delete")).toBeInTheDocument();

        const closeButton = screen.getByText("X");
        userEvent.click(closeButton);
        expect(screen.queryByTestId("context-menu")).not.toBeInTheDocument();
    });

    it("handle click on the movie tile image", () => {
        const onClick = jest.fn();
        render(<MovieTile movieDetails={mockMoviesData}
                          onClick={onClick}
                          handleDelete={() => {}}
                          handleEdit={() => {}} />);
        userEvent.click(screen.getByRole("img"));
        expect(onClick).toBeCalled();
    });

    it("renders with fallback image on error", () => {
        render(<MovieTile
            movieDetails={mockMoviesData}
            onClick={() => {}}
            handleDelete={() => {}}
            handleEdit={() => {}} />);

        const image = screen.getByAltText(mockMoviesData.title);
        expect(image).toHaveAttribute("src", mockMoviesData.poster_path);
        fireEvent.error(image);
        expect(image).toHaveAttribute("src", placeholderImage);
    });

    it("renders with fallback image on null", () => {
        const mockMoviesDataWithNullPoster = {
            ...mockMoviesData,
            poster_path: null,
        }
        render(<MovieTile
            movieDetails={mockMoviesDataWithNullPoster}
            onClick={() => {}}
            handleDelete={() => {}}
            handleEdit={() => {}} />);

        const image = screen.getByAltText(mockMoviesData.title);
        expect(image).toHaveAttribute("src", placeholderImage);
    });

})
