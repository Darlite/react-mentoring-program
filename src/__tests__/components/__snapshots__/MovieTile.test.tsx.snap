// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieTile renders correctly 1`] = `
<div>
  <div
    class="movieTile"
    data-testid="movie-tile"
  >
    <img
      alt="Pulp Fiction"
      class="movieTileImage"
      src="images/Pulp_Fiction.png"
    />
    <span
      class="movieTitleAndYear"
    >
      <h2
        class="title"
      >
        Pulp Fiction
      </h2>
      <span
        class="year"
        data-cy="movie-tile-release-date"
      >
        1994
      </span>
    </span>
    <p
      class="genres"
      data-cy="movie-tile-genres"
    >
      <span>
        Crime
         
      </span>
      <span>
        Thriller
         
      </span>
    </p>
  </div>
</div>
`;
